### PR TITLE
✨ Implement Call and CallAll in RuntimeClient

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:

Implements the `Call` and `CallAll` functions in the RuntimeClient. These functions are used to make the http calls to the Runtime Extension Handlers.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

Part of #6330

/kind feature
/kind runtime-sdk

### CODE REVIEW

File: internal/runtime/client/client.go [hunk 1]
```diff
@@ -153,6 +157,106 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
+	gvh, err := h.client.catalog.GroupVersionHook(h.hook)
+	if err != nil {
+		return errors.Wrap(err, "failed to compute GroupVersionHook")
+	}
+	registrations, err := h.client.registry.List(catalog.GroupHook{Group: gvh.Group, Hook: gvh.Hook})
+	if err != nil {
+		return errors.Wrap(err, "failed to retrieve ExtensionHandlers information")
+	}
+	responses := []*runtimehooksv1.ExtensionHandlerResponse{}
+	// Future-work: Call the extension handlers concurrently.
```
Comment (sbueringer) [inline internal/runtime/client/client.go:191]: I would suggest to drop this comment entirely for now. Let's just see if that requirement comes up later. I'm not sure if that optimization is needed.

File: internal/runtime/client/client.go [hunk 2]
```diff
@@ -153,6 +157,106 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
+	if err := response.Aggregate(responses); err != nil {
+		return errors.Wrap(err, "failed to aggregate responses")
+	}
+	return nil
+}
+
+func (h *hookClient) Call(ctx context.Context, name string, request, response runtime.Object) error {
+	registration, err := h.client.registry.Get(name)
+	if err != nil {
+		return err
```
Comment (sbueringer) [inline internal/runtime/client/client.go:216]: nit: error wrapping for more context would be good

File: internal/runtime/client/client.go [hunk 3]
```diff
@@ -153,6 +157,106 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
+		timeout: timeoutDuration,
+	}
+	err = httpCall(ctx, request, response, opts)
+	if err != nil {
+		// If the error is errCallingExtensionHandler then apply failure policy to calculate
+		// the effective result of the operation.
+		ignore := *registration.FailurePolicy == runtimev1.FailurePolicyIgnore
+		if _, ok := err.(*errCallingExtensionHandler); ok && ignore {
+			// Update the response to a default success response and return.
+			if err := runtimehooksv1.SetStatus(response, runtimehooksv1.ResponseStatusSuccess); err != nil {
```
Comment (sbueringer) [inline internal/runtime/client/client.go:236]: I wonder if we should also set the message to empty at this point, because it is supposed to contain an error message

File: internal/runtime/client/client.go [hunk 4]
```diff
@@ -250,12 +354,18 @@ func httpCall(ctx context.Context, request, response runtime.Object, opts *httpC
...
 	resp, err := client.Do(httpRequest)
 	if err != nil {
-		return errors.Wrap(err, "failed to perform the http call")
+		return &errCallingExtensionHandler{
+			extensionHandlerName: opts.name,
+			err:                  errors.Wrap(err, "failed to make the http call"),
+		}
 	}
 
 	defer resp.Body.Close()
```
Comment (sbueringer) [inline internal/runtime/client/client.go:363]: just a reminder. I think we wanted to check for != 200 here (but not sure)

File: internal/runtime/client/client.go [hunk 5]
```diff
@@ -302,3 +412,13 @@ func urlForExtension(config runtimev1.ClientConfig, gvh catalog.GroupVersionHook
...
 }
+
+type errCallingExtensionHandler struct {
+	extensionHandlerName string
+	err                  error
+}
+
+func (e *errCallingExtensionHandler) Error() string {
+	// FIXME: find a better error message.
+	return fmt.Sprintf("failed processing handler %s with error: %s", e.extensionHandlerName, e.err)
```
Comment (sbueringer) [inline internal/runtime/client/client.go:423]: What about
```suggestion
	return fmt.Sprintf("failed to call Extensionhandler %q: %v", e.extensionHandlerName, e.err)
```

**Review by sbueringer — COMMENTED** [review]
Just a high-level review, looks good overall

Just let me know when I should take a closer look

File: internal/runtime/client/client.go (hunk 5 above)
Comment (ykakarap) [reply internal/runtime/client/client.go:423]: The same error will be returned for network error, non 200 response objects and response decoding errors. Not sure if "failed to call...." is the right wording for all of them.

File: internal/runtime/client/client.go [hunk 6]
```diff
@@ -75,6 +76,12 @@ type Client interface {
 
 	// Unregister unregisters the ExtensionConfig.
 	Unregister(extensionConfig *runtimev1.ExtensionConfig) error
+
+	// CallAll calls all the extension registered for the hook.
+	CallAll(ctx context.Context, hook catalog.Hook, request runtime.Object, response runtimehooksv1.AggregatableResponse) error
+
+	// Call calls only the extension with the given name.
+	Call(ctx context.Context, name string, request, response runtime.Object) error
```
Comment (ykakarap) [inline internal/runtime/client/client.go:84]: How about 
```go
// CallHook calls all the extensions registered for the hook.
CallHook(ctx context.Context, hook catalog.Hook, request runtime.Object, response runtimehooksv1.AggregatableResponse) error

// CallExtension calls the extension with the given name.
CallExtension(ctx context.Context, name string, request, response runtime.Object) error
```

How about using `CallHook` and `CallExtension` or (`CallByHook` and `CallByExtension`) instead of `CallAll` and `Call`?

Once we remove the builder pattern the usage is going to be `runtimeClient.CallAll()` which seems unclear and ambiguous.

**Comment by ykakarap** [comment]
@sbueringer I addressed the initial round of feedback. This is ready for a full review.

File: exp/runtime/hooks/api/v1alpha1/util.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+func getField(obj interface{}, field string) (interface{}, error) {
+	v, err := valueOf(obj)
+	if err != nil {
+		return "", err
+	}
+	f := v.FieldByName(field)
+	if !f.IsValid() {
+		return "", fmt.Errorf("unexpected type. filed %s does not exist", field)
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/util.go:100]:
```suggestion
		return "", fmt.Errorf("unexpected type. field %s does not exist", field)
```

File: exp/runtime/hooks/api/v1alpha1/util_test.go [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+func TestGetMessage(t *testing.T) {
+	tests := []struct {
+		name    string
+		obj     *DiscoveryResponse
+		want    string
+		wantErr bool
+	}{
+		{
+			name: "get the message field from the object",
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/util_test.go:110]: Please add one test cases with invalidResponseObject like above + one test case with an empty message

File: exp/runtime/hooks/api/v1alpha1/util_test.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+
+func TestSetMessage(t *testing.T) {
+	tests := []struct {
+		name    string
+		obj     *DiscoveryResponse
+		message string
+		wantErr bool
+	}{
+		{
+			name:    "setting the status should succeeded",
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/util_test.go:141]: let's add one test case to overwrite an already existing message

File: exp/runtime/hooks/api/v1alpha1/util_test.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+			},
+			want:    ResponseStatusSuccess,
+			wantErr: false,
+		},
+		{
+			name: "get the status field from the object - failure",
+			obj: &DiscoveryResponse{
+				Status: ResponseStatusFailure,
+			},
+			want:    ResponseStatusFailure,
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/util_test.go:181]: let's add the invalidResponseObject test cases as well

File: exp/runtime/hooks/api/v1alpha1/util_test.go [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+func TestSetStatus(t *testing.T) {
+	tests := []struct {
+		name    string
+		obj     *DiscoveryResponse
+		status  ResponseStatus
+		wantErr bool
+	}{
+		{
+			name:    "setting the status should succeeded",
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/util_test.go:208]: let's add one overwrite test case

File: internal/runtime/client/client.go (hunk 6 above)
Comment (sbueringer) [reply internal/runtime/client/client.go:84]: I would suggest: `CallExtensions` (or `CallAllExtensions`) and `CallExtension`.

I would like to avoid mixing call hook and call extension, as a user I would wonder what the difference between a hook and an extension is.

I wonder if it would be better if we also add catalog.Hook as a parameter to `CallExtension`. I know we don't strictly need it, but it might be better if a user has to explicitly declare what hook they want to call and we use that information to make sure that the type matches what is in the registry (i.e. registration.GroupVersionHook).

@fabriziopandini We discussed it and for now we would like to avoid mixing top-level discovery calls and extension calls on a deeper layer. I think it's better to keep it all flat at this point and not mix patterns and then potentially change it when we reconsider this later.

File: internal/runtime/client/client.go [hunk 12]
```diff
@@ -153,6 +160,89 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
 }
 
+func (c *client) CallAll(ctx context.Context, hook catalog.Hook, request runtime.Object, response runtimehooksv1.AggregatableResponse) error {
+	gvh, err := c.catalog.GroupVersionHook(hook)
+	if err != nil {
+		return errors.Wrap(err, "failed to compute GroupVersionHook")
+	}
+	registrations, err := c.registry.List(catalog.GroupHook{Group: gvh.Group, Hook: gvh.Hook})
+	if err != nil {
+		return errors.Wrap(err, "failed to retrieve ExtensionHandlers information")
```
Comment (sbueringer) [inline internal/runtime/client/client.go:170]:
```suggestion
		return errors.Wrap(err, "failed to retrieve ExtensionHandlers with group %q and hook %q", gvh.Group, gvh.Hook)
```
nit

File: internal/runtime/client/client.go [hunk 13]
```diff
@@ -153,6 +160,89 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
+	}
+	responses := []*runtimehooksv1.ExtensionHandlerResponse{}
+	for _, registration := range registrations {
+		tmpResponse, err := c.catalog.NewResponse(gvh)
+		if err != nil {
+			return errors.Wrap(err, "failed to create respons object")
```
Comment (sbueringer) [inline internal/runtime/client/client.go:176]:
```suggestion
			return errors.Wrap(err, "failed to create response object")
```
nit

File: internal/runtime/client/client.go [hunk 14]
```diff
@@ -153,6 +160,89 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
+	if err := response.Aggregate(responses); err != nil {
+		return errors.Wrap(err, "failed to aggregate responses")
+	}
+	return nil
+}
+
+func (c *client) Call(ctx context.Context, name string, request, response runtime.Object) error {
+	registration, err := c.registry.Get(name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to retrieve registration information for %s", name)
```
Comment (sbueringer) [inline internal/runtime/client/client.go:197]:
```suggestion
		return errors.Wrapf(err, "failed to retrieve ExtensionHandler with name %q", name)
```

File: internal/runtime/client/client.go [hunk 15]
```diff
@@ -153,6 +160,89 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
 	return nil
 }
 
+func (c *client) CallAll(ctx context.Context, hook catalog.Hook, request runtime.Object, response runtimehooksv1.AggregatableResponse) error {
```
Comment (sbueringer) [inline internal/runtime/client/client.go:163]: Please add the godoc comments from the interface here and on Call as well.
I think we also need some documentation about under which circumstances we return an error and how we modify the response based on failurePolicy, ... (for both Call and CallAll)

File: internal/runtime/client/client.go [hunk 16]
```diff
@@ -250,12 +340,25 @@ func httpCall(ctx context.Context, request, response runtime.Object, opts *httpC
...
+		return &errCallingExtensionHandler{
+			extensionHandlerName: opts.name,
+			err:                  errors.Wrap(err, "failed to make the http call"),
+		}
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return &errCallingExtensionHandler{
+			extensionHandlerName: opts.name,
+			err:                  fmt.Errorf("non 200 response status received"),
```
Comment (sbueringer) [inline internal/runtime/client/client.go:352]:
```suggestion
			err:                  errors.Errorf("non 200 response status received"),
```
I think we should use this lib consistently to get consistent errors. Let's please add the status code we got to the error message

File: internal/runtime/client/client.go (hunk 5 above)
Comment (sbueringer) [reply internal/runtime/client/client.go:416]: Then the name of the struct is not ideal :)

What about simplifying this. We can make this custom type just:
```
type errCallingExtensionHandler error
```
and then we can write the exact right error message when we return. I think making those errors detectable one layer above does not necessarily mean we need a common error message style, e.g.
```
return errCallingExtensionHandler(errors.Wrapf(err, "failed to call ExtensionHandler: %q", opts.name))
```
(or if we have to keep the struct because we need access to `extensionHandlerName` we can just return `e.err.Error()`).

File: internal/runtime/client/client.go [hunk 17]
```diff
@@ -153,6 +160,89 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
 	return nil
 }
```
Comment (sbueringer) [inline internal/runtime/client/client.go:162]: Let's please add some unit tests for Call and CallAll. If I see correctly it should be relatively straightforward (what we've done for httpCall + we have to setup a registry (by just registering stuff and skipping discovery)

File: exp/runtime/hooks/api/v1alpha1/aggregate.go [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply exp/runtime/hooks/api/v1alpha1/aggregate.go:22]: Note: I discussed it with Yuvaraj and we agree that we'll change it in a follow-up PR, but for now we keep it as is.
This is also the reason why we won't invest too much into the reflection unit tests for now, with the introduction of common structs we are probably able to drop the reflection code altogether. (and if we don't, we will add more unit testing for the reflection code then)

File: internal/runtime/client/client.go [hunk 19]
```diff
@@ -250,12 +365,22 @@ func httpCall(ctx context.Context, request, response runtime.Object, opts *httpC
...
 	if err != nil {
-		return errors.Wrap(err, "failed to perform the http call")
+		return errCallingExtensionHandler(
+			errors.Wrapf(err, "failed to call ExtensionHandler: %q", opts.name),
+		)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return errCallingExtensionHandler(
+			fmt.Errorf("non 200 response code, %q, not accepted", resp.StatusCode),
```
Comment (sbueringer) [inline internal/runtime/client/client.go:375]:
```suggestion
			errors.Errorf("non 200 response code, %q, not accepted", resp.StatusCode),
```
nit (for consistent library usage)

File: internal/runtime/client/client_test.go [hunk 20]
```diff
@@ -269,3 +273,279 @@ func TestURLForExtension(t *testing.T) {
...
+			args: args{
+				hook:     fakev1alpha1.FakeHook,
+				name:     "unregistered-extension",
+				request:  &fakev1alpha1.FakeRequest{},
+				response: &fakev1alpha1.FakeResponse{},
+			},
+			wantErr: true,
+		},
+		{
+			name:                   "should when extension handler is not compatible with the hook",
```
Comment (sbueringer) [inline internal/runtime/client/client_test.go:387]:
```suggestion
			name:                   "should fail when extension handler from the registry is not compatible with the hook arg",
```

File: internal/runtime/client/client_test.go [hunk 21]
```diff
@@ -269,3 +273,279 @@ func TestURLForExtension(t *testing.T) {
...
+
+func TestClient_CallExtension(t *testing.T) {
+	g := NewWithT(t)
+
+	testHostPort := "127.0.0.1:9090"
+
+	fpFail := runtimev1.FailurePolicyFail
+	fpIgnore := runtimev1.FailurePolicyIgnore
+
+	validExtensionHandlerWithFailPolicy := runtimev1.ExtensionConfig{
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply internal/runtime/client/client_test.go:285]: Note: if creating an ExtensionConfig is ~ a one liner, should we change the test case parameter to `registeredExtensionHandlers` and create the registry on the fly in the for loop? I think this would be easier as there is more useful info directly visible in the table test case

**Review by killianmuldoon — COMMENTED** [review]
Also  - I can't resolve conversations on this PR :laughing:

File: internal/runtime/client/client.go [hunk 22]
```diff
@@ -153,6 +163,107 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
+		responses = append(responses, response)
+	}
+	if err := aggregateResponses(responses); err != nil {
+		return errors.Wrap(err, "failed to aggregate responses")
+	}
+	return nil
+}
+
+func aggregateResponses(responses []runtimehooksv1.Response) error {
+	// TODO:(killianmuldoon) implement proper aggregation logic.
```
Comment (sbueringer) [inline internal/runtime/client/client.go:204]: Let's make this a follow-up PR with handling for RetryableResponse, unit tests etc

File: internal/runtime/client/client.go [hunk 23]
```diff
@@ -153,6 +163,107 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
+	}
+	err = httpCall(ctx, request, response, opts)
+	if err != nil {
+		// If the error is errCallingExtensionHandler then apply failure policy to calculate
+		// the effective result of the operation.
+		ignore := *registration.FailurePolicy == runtimev1.FailurePolicyIgnore
+		if _, ok := err.(errCallingExtensionHandler); ok && ignore {
+			// Update the response to a default success response and return.
+			// - Set status to success
+			// - Set message to empty string
```
Comment (sbueringer) [inline internal/runtime/client/client.go:251]:
```suggestion
```
nit, would drop those comments to make it a bit shorter given that it's almost the same as the actual code

File: internal/runtime/client/client.go [hunk 24]
```diff
@@ -153,6 +163,107 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
+			response.SetStatus(runtimehooksv1.ResponseStatusSuccess)
+			response.SetMessage("")
+			return nil
+		}
+		return errors.Wrap(err, "failed to call extension")
```
Comment (sbueringer) [inline internal/runtime/client/client.go:256]:
```suggestion
		return errors.Wrap(err, "failed to call ExtensionHandler")
```

also use upper case in l.260. It's fine for names. lower case is just that the error concat with `:` looks nicer

**Comment by sbueringer** [comment]
> Also - I can't resolve conversations on this PR 😆

Just :+1: on the conversations you would resolve and I'll do it

File: internal/runtime/catalog/test/catalog_test.go [hunk 25]
```diff
@@ -0,0 +1,282 @@
...
+	}
+}
+
+type GoodRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	First string `json:"first"`
+}
+
+func (in *GoodRequest) DeepCopyObject() runtime.Object {
```
Comment (sbueringer) [inline internal/runtime/catalog/test/catalog_test.go:184]: Q: What do you think about generating the deepcopy stuff for the fake types the same way that we generate conversion? It's probably simpler to adjust the Makefile a bit vs. adding the funcs over time with panics :)

File: internal/runtime/test/v1alpha1/fake_types.go [hunk 26]
```diff
@@ -72,6 +72,10 @@ type FakeResponse struct {
 	First  int
 }
 
+func (in *FakeResponse) DeepCopyObject() runtime.Object {
```
Comment (sbueringer) [inline internal/runtime/test/v1alpha1/fake_types.go:75]: Q: What do you think about generating the deepcopy stuff for the fake types the same way that we generate conversion? It's probably simpler to adjust the Makefile a bit vs. adding the funcs over time with panics :) 

If I see correctly:
```
		paths=./internal/runtime/test/v1alpha1 \
		paths=./internal/runtime/test/v1alpha2 \
```
to generate-go-deepcopy-core

File: exp/runtime/hooks/api/v1alpha1/common_types.go [hunk 27]
```diff
@@ -16,6 +16,46 @@ limitations under the License.
 
 package v1alpha1
 
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Response can set and get fields common to all Response types.
+// +kubebuilder:object:generate=false
+type Response interface {
```
Comment (fabriziopandini) [inline exp/runtime/hooks/api/v1alpha1/common_types.go:25]: Q: should this be `ReponseObject` given that it is an object + some behavior that are specific for a ResponseObject?

File: exp/runtime/hooks/api/v1alpha1/common_types.go [hunk 28]
```diff
@@ -16,6 +16,46 @@ limitations under the License.
 
 package v1alpha1
 
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Response can set and get fields common to all Response types.
```
Comment (fabriziopandini) [inline exp/runtime/hooks/api/v1alpha1/common_types.go:23]:
```suggestion
// ResponseObject is a runtime object extended with methods to handle response-specific fields.
```

File: exp/runtime/hooks/api/v1alpha1/common_types.go [hunk 29]
```diff
@@ -16,6 +16,46 @@ limitations under the License.
...
+type Response interface {
+	runtime.Object
+	GetMessage() string
+	GetStatus() ResponseStatus
+	SetMessage(message string)
+	SetStatus(status ResponseStatus)
+}
+
+// CommonResponse is the data structure common to all Response types.
+type CommonResponse struct {
```
Comment (fabriziopandini) [inline exp/runtime/hooks/api/v1alpha1/common_types.go:34]: What about
```suggestion
// ResponseMeta describes metadata that all response messages must have, including status
// and error message.
type ResponseMeta struct {
```

File: internal/runtime/client/client.go [hunk 30]
```diff
@@ -153,6 +163,98 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
+	gvh, err := c.catalog.GroupVersionHook(hook)
+	if err != nil {
+		return errors.Wrap(err, "failed to compute GroupVersionHook")
+	}
+	registration, err := c.registry.Get(name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to retrieve ExtensionHandler with name %q", name)
+	}
+	if gvh.GroupHook() != registration.GroupVersionHook.GroupHook() {
+		return fmt.Errorf("ExtensionHandler %q does not match group %s, hook %s", name, gvh.Group, gvh.Hook)
```
Comment (fabriziopandini) [inline internal/runtime/client/client.go:223]: nit: Let's use consistently `github.com/pkg/errors`
(note this applies to multiple places)

File: internal/runtime/client/client.go [hunk 31]
```diff
@@ -153,6 +163,98 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
+	err = httpCall(ctx, request, response, opts)
+	if err != nil {
+		// If the error is errCallingExtensionHandler then apply failure policy to calculate
+		// the effective result of the operation.
+		ignore := *registration.FailurePolicy == runtimev1.FailurePolicyIgnore
+		if _, ok := err.(errCallingExtensionHandler); ok && ignore {
+			// Update the response to a default success response and return.
+			response.SetStatus(runtimehooksv1.ResponseStatusSuccess)
+			response.SetMessage("")
+			return nil
```
Comment (fabriziopandini) [inline internal/runtime/client/client.go:245]: nit: we should log something when this happens

File: internal/runtime/client/client.go [hunk 32]
```diff
@@ -153,6 +163,98 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
+func aggregateResponses(responses []runtimehooksv1.Response) error {
+	// TODO:(killianmuldoon) implement proper aggregation logic.
+	return nil
+}
+
+// CallExtension make the call to the extension with the given name.
+// The response object passed will be updated with the response of the call.
+// An error is returned if the extension is not compatible with the hook.
+// If the ExtensionHandler returns a response with `Status` set to `Failure` the function returns an error
+// and the response object is updated with the response received from the extension handler.
```
Comment (fabriziopandini) [inline internal/runtime/client/client.go:205]: I'm not sure I can find this behavior in the code below 😓

File: internal/runtime/client/client.go [hunk 33]
```diff
@@ -153,6 +163,98 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
...
+//
+// FailurePolicy of the ExtensionHandler is used to handle errors that occur when performing the external call to the extension.
+// - If FailurePolicy is set to Ignore, the error is ignored and the response object is updated to be the default success response.
+// - If FailurePolicy is set to Fail, an error is returned and the response object may or may not be updated.
+// Nb. FailurePolicy does not affect the following kinds of errors:
+// - Internal errors. Examples: hooks is incompatible with ExtensionHandler, ExtensionHandler information is missing.
+// - Error when ExtensionHandler returns a response with `Status` set to `Failure`.
```
Comment (fabriziopandini) [inline internal/runtime/client/client.go:212]: I think we should surface this in the also in the API where FailurePolicy is defined.
We should also add an explanation about how we are interpreting failure policy, which is consistent with Admission webhooks; e.g.
Failure policy helps with extensions not working consistently; however, the following type of errors are always considered blocking Failures
- misconfigurations (incompatible types) 
- when the extension explicitly report a Status Failure

File: internal/runtime/client/client.go [hunk 34]
```diff
@@ -153,6 +163,98 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
 	return nil
 }
 
+// CallAllExtensions calls all the ExtensionHandlers registered for the hook.
+// The ExtensionHandler are called sequentially. The function exits immediately after any of the ExtensionHandlers return an error.
```
Comment (fabriziopandini) [inline internal/runtime/client/client.go:167]: I'm not sure about this behavior. Is this derived/consistent with admission webhooks?

**Review by fabriziopandini — COMMENTED** [review]
Some nits and questions.
Also, as soon as we are in a state where we can test all the pieces working together we should add some logging to call and callAll

File: exp/runtime/hooks/api/v1alpha1/common_types.go (hunk 29 above)
Comment (sbueringer) [reply exp/runtime/hooks/api/v1alpha1/common_types.go:34]: Is this really metadata?

File: internal/runtime/client/client.go (hunk 31 above)
Comment (killianmuldoon) [reply internal/runtime/client/client.go:245]: Is it alright to push logging to a follow up PR? It would be good to handle it holistically from a whole-system POV.

File: internal/runtime/client/client.go (hunk 34 above)
Comment (sbueringer) [reply internal/runtime/client/client.go:167]: I don't remember, but @ykakarap researched it.

But I'm pretty sure we can't be entirely consistent with how admission webhooks work today, as it's e.g. possible that mutating webhooks are called multiple times during a single admission IIRC.

Comment (sbueringer) [reply internal/runtime/client/client.go:167]: We had some discussion about always calling all of them (which can take very long if multiple are broken and timeout) or calling them in parallel, but for now we wanted to keep it as simple as possible.

File: internal/runtime/client/client.go (hunk 32 above)
Comment (killianmuldoon) [reply internal/runtime/client/client.go:205]: We return the error in L:251 is my understanding, and the response is mutated inside the httpCall method L:374. Maybe I'm misunderstanding the code or the comment though :neutral_face:

Comment (sbueringer) [reply internal/runtime/client/client.go:205]: I think Killian is correct

File: exp/runtime/hooks/api/v1alpha1/common_types.go (hunk 29 above)
Comment (sbueringer) [reply exp/runtime/hooks/api/v1alpha1/common_types.go]: I'm also not sure how we can build on top of this naming when we add another struct to cover the RetryAfterSeconds field.

File: internal/runtime/client/client.go (hunk 34 above)
Comment (fabriziopandini) [reply internal/runtime/client/client.go:167]: What is important for me is that we document WHY are we implementing a certain behavior, otherwise in future we will circle back to the same discussion

File: exp/runtime/hooks/api/v1alpha1/common_types.go (hunk 29 above)
Comment (sbueringer) [reply exp/runtime/hooks/api/v1alpha1/common_types.go]: We would prefer CommonResponse as those fields feel more like fields common to Responses.

They only tell us if the response as a whole is success/failure, but apart from that they don't tell us anything about other attributes of the response (so meta doesn't seem to fit well)

**Review by sbueringer — COMMENTED** [review]
~ last round of nits from my side

File: Makefile [hunk 35]
```diff
@@ -257,6 +257,7 @@ generate-go-deepcopy-core: $(CONTROLLER_GEN) ## Generate deepcopy go code for co
 		paths=./$(EXP_DIR)/addons/api/... \
 		paths=./$(EXP_DIR)/runtime/api/... \
 		paths=./$(EXP_DIR)/runtime/hooks/api/... \
+        paths=./internal/runtime/test/... \
```
Comment (chrischdi) [inline Makefile:260]:
```suggestion
		paths=./internal/runtime/test/... \
```

**Comment by sbueringer** [comment]
Thank you very much

lgtm pending squash from my side!

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve
